### PR TITLE
feat: Introduce CDK 🎉 

### PR DESCRIPTION
**Pull request by akash1810**
Builds on #157.

It should be easier to review this PR commit by commit.

## What does this change?

In this change we introduce an empty GuCDK project via:

```sh
npx @guardian/cdk@latest new \
  --app cloudwatch-logs-management \
  --stage PROD \
  --stack cloudwatch-logs-management \
  --package-manager npm
```

Unfortunately, the existing TypeScript project wasn't very happy with this, stating:

```log
error TS6059: File '/Users/akash_askoolum/code/cloudwatch-logs-management/cdk/bin/cdk.ts' is not under 'rootDir' '/Users/akash_askoolum/code/cloudwatch-logs-management/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by include pattern '**/*' in '/Users/akash_askoolum/code/cloudwatch-logs-management/tsconfig.json'

error TS6059: File '/Users/akash_askoolum/code/cloudwatch-logs-management/cdk/lib/cloudwatch-logs-management.test.ts' is not under 'rootDir' '/Users/akash_askoolum/code/cloudwatch-logs-management/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by include pattern '**/*' in '/Users/akash_askoolum/code/cloudwatch-logs-management/tsconfig.json'

Found 2 errors.
```

I took the chance to adopt [NPM workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces) to resolve this. This makes up the majority of the changes in this PR.

Workspaces provide the added benefit of de-duping dependencies (both the app and cdk require typescript, eslint, etc). Ditto for the configuration of these dependencies.

### Other things worth noting
The CDK project on this branch doesn't actually do anything; it doesn't take ownership of the YAML template, or even have any resources defined. We're just wiring it into the CI/CD process for now.

## What testing has been performed for this change?

I've deployed this to the DevX account, and can continue to see logs. This can be confirmed by the value of the `ShippedBy-revision` marker on the logs - it matches the hash of the last commit on this branch.

image

## How can we measure success?

We open the gates to a full CDK migration!

## Have we considered potential risks?

Are workspaces confusing?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,24 +1,23 @@
 {
-  "name": "cloudwatch-logs-management",
-  "version": "1.0.0",
+  "name": "@guardian/cloudwatch-logs-management",
```
Comment (akash1810) [inline package.json:2]: The `package.json` at the root level contains all common dependencies, e.g. typescript, eslint, prettier.

File: packages/app/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "private": true,
+  "version": "0.0.0",
+  "description": "Lambdas to maintain and forward cloudwatch logs",
+  "scripts": {
+    "build": "tsc",
+    "test": "jest"
+  },
+  "devDependencies": {
+    "@types/aws-lambda": "^8.10.20",
+    "aws-sdk": "^2.1159.0"
```
Comment (akash1810) [inline packages/app/package.json:12]: Only the app uses these dependencies, so are included in `packages/app/package.json` _not_ the root `package.json`.

File: packages/app/src/logEntryProcessing.ts [hunk 3]
```diff
@@ -111,6 +111,9 @@ export function lambdaRequestLogData(
 					lambdaStats: stats,
 				};
 			}
+			default: {
+				return undefined;
+			}
```
Comment (akash1810) [inline packages/app/src/logEntryProcessing.ts:116]: The `tsconfig.json` at the root is now quite strict. Changes to this file resolves these errors:

```
error TS7030: Not all code paths return a value.
```

File: packages/cdk/.gitignore [hunk 4]
```diff
@@ -0,0 +1,11 @@
+*.js
```
Comment (akash1810) [inline packages/cdk/.gitignore:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Infrastructure
```
Comment (akash1810) [inline packages/cdk/README.md:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/bin/cdk.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import 'source-map-support/register';
```
Comment (akash1810) [inline packages/cdk/bin/cdk.ts:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/cdk.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (akash1810) [inline packages/cdk/cdk.json:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/jest.setup.js [hunk 8]
```diff
@@ -0,0 +1 @@
+jest.mock("@guardian/cdk/lib/constants/tracking-tag");
```
Comment (akash1810) [inline packages/cdk/jest.setup.js:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/lib/__snapshots__/cloudwatch-logs-management.test.ts.snap [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (akash1810) [inline packages/cdk/lib/__snapshots__/cloudwatch-logs-management.test.ts.snap:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/lib/cloudwatch-logs-management.test.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { App } from 'aws-cdk-lib';
```
Comment (akash1810) [inline packages/cdk/lib/cloudwatch-logs-management.test.ts:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/lib/cloudwatch-logs-management.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { join } from 'path';
```
Comment (akash1810) [inline packages/cdk/lib/cloudwatch-logs-management.ts:1]: Provided by the GuCDK project generator. Can be ignored.

File: packages/cdk/package.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    "synth": "cdk synth --path-metadata false --version-reporting false",
+    "diff": "cdk diff --path-metadata false --version-reporting false"
+  },
+  "devDependencies": {
+    "@guardian/cdk": "45.1.3",
+    "aws-cdk": "2.25.0",
+    "aws-cdk-lib": "2.25.0",
+    "constructs": "10.1.17",
+    "source-map-support": "^0.5.20",
+    "ts-node": "^10.8.1"
```
Comment (akash1810) [inline packages/cdk/package.json:17]: Only cdk uses these dependencies, so are included in packages/cdk/package.json not the root package.json.

File: scripts/ci.sh [hunk 13]
```diff
@@ -16,15 +16,16 @@ setupNodeVersion() {
...
 }
 
 setupNodeVersion
 injectBuildInfo
 
 npm ci
-npm run test
 npm run lint
-npm run build
+npm run test --workspaces
```
Comment (akash1810) [inline scripts/ci.sh:28]: Run the `test` script from _every_ workspace.

File: tsconfig.json [hunk 14]
```diff
@@ -1,13 +1,31 @@
...
+        "noImplicitThis": true,
+        "alwaysStrict": true,
+        "noUnusedLocals": false,
+        "noUnusedParameters": false,
+        "noImplicitReturns": true,
+        "noFallthroughCasesInSwitch": false,
+        "inlineSourceMap": true,
+        "inlineSources": true,
+        "experimentalDecorators": true,
+        "strictPropertyInitialization": false,
```
Comment (akash1810) [inline tsconfig.json:26]: Super strict! This was copied from the `tsconfig.json` file provided by the GuCDK project generator.

**Comment by akash1810** [comment]
cc @jfsoul as usage of NPM workspaces might impact Snyk coverage on this repo?

**Review by jacobwinch — APPROVED** [review]
> The CDK project on this branch doesn't actually do anything; it doesn't take ownership of the YAML template, or even have any resources defined. We're just wiring it into the CI/CD process for now.

Thanks for breaking this risky change down; this (along with the clear commit history) makes it much easier for reviewers 💯 

Great work 👍

**Comment by jacobwinch** [comment]
> usage of NPM workspaces might impact Snyk coverage on this repo?

I can't 100% remember the compatibility status of this. Let's merge as is (to avoid rebase hell) and swap to yarn workspaces as a follow-up, if necessary?

**Pull request merged**